tutorial/part11.md

```diff
@@ -0,0 +1,994 @@
+# Part 11 - Delving into the Dungeon
+
+Our dungeon lacks depth. There is afterall only a single floor to explore. In this part we will again refactor things - this time to support z levels, and to generate dungeon floors on the fly. We will have persistant randomly generated dungeon floors with stairs to link them. That's a lot to do so let's get started!
+
+## Supporting z levels in the cache
+
+Our cache will need to store the current z level and some details about each floor that has been created. This is how we will be able to generate persistent levels on the way down that we can revisit on the way back up.
+
+In `./src/state/cache.js`:
+
+```diff
++import { get, set } from "lodash";
++
+export const cache = {
+  entitiesAtLocation: {},
++  z: -1,
++  floors: {}, // { z: { stairsUp: '', stairsDown: '' } }
+};
+
++export const addCache = (path, value) => {
++  set(cache, path, value);
++};
++
++export const readCache = (path) => get(cache, path);
++
+export const addCacheSet = (name, key, value) => {
+  if (cache[name][key]) {
+    cache[name][key].add(value);
+```
+
+```diff
+export const serializeCache = () => {
+  const entitiesAtLocation = Object.keys(cache.entitiesAtLocation).reduce(
+    (acc, val) => {
+      acc[val] = [...cache.entitiesAtLocation[val]];
+      return acc;
+    },
+    {}
+  );
+
+  return {
+    entitiesAtLocation,
++    z: cache.z,
++    floors: cache.floors,
+  };
+};
+```
+
+```diff
+export const deserializeCache = (data) => {
+  cache.entitiesAtLocation = Object.keys(data.entitiesAtLocation).reduce(
+    (acc, val) => {
+      acc[val] = new Set(data.entitiesAtLocation[val]);
+      return acc;
+    },
+    {}
+  );
+
++  cache.z = data.z;
++  cache.floors = data.floors;
+};
+
+export const clearCache = () => {
+  cache.entitiesAtLocation = {};
++  cache.z = 1;
+  cache.floors = {};
+};
+```
+
+Before moving on the next section, test your game and make sure everything is still working exactly as it was before.
+
+## Refactor Z
+
+Our cache now supports z levels but the rest of our game does not. That is going to require a pretty hefty refactor. While it would have been nice if we had anticipated this need from the get go - I would argue it's best to code for what you need at the time, not what you might want down the line. Refactoring is just a natural part of the gig. There are 10 files to adjust and for clarity, I have each file as a heading with the changes below it. There's no commentary for the rest of this section. Just grab a cup of coffee or tea and get on it.
+
+### `./src/index.js`
+
+```diff
+import {
+  clearCache,
+  deserializeCache,
++  readCache,
+  readCacheSet,
+  serializeCache,
+} from "./state/cache";
+```
+
+```diff
+const loadGame = () => {
+  const data = JSON.parse(localStorage.getItem("gameSaveData"));
+  if (!data) {
+    addLog("Failed to load - no saved games found");
+    return;
+  }
+
+  for (let entity of ecs.entities.all) {
+    entity.destroy();
+  }
```

Comment:
this would be a nice feature for the library. to be able to "clear" all entities